src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/ServerSentEventsMessageFormatter.cs

```diff
@@ -12,113 +12,101 @@ namespace Microsoft.AspNetCore.Http.Connections.Internal
+        /// Gets the last memory segment in a sequence.
+        /// </summary>
+        /// <param name="source">Source sequence.</param>
+        /// <param name="offset">The offset the segment starts at.</param>
+        /// <returns>The last memory segment in a sequence.</returns>
+        private static ReadOnlyMemory<byte> GetLastSegment(in ReadOnlySequence<byte> source, out long offset)
         {
-            // Write the payload
-            WritePayload(payload, output);
-
-            // Write new \r\n
-            output.Write(Newline, 0, Newline.Length);
-        }
+            offset = 0;
 
-        private static void WritePayload(ReadOnlyMemory<byte> payload, Stream output)
-        {
-            // Short-cut for empty payload
-            if (payload.Length == 0)
+            var totalLength = source.Length;
+            var position = source.Start;
+            while (source.TryGet(ref position, out ReadOnlyMemory<byte> memory))
             {
-                return;
+                // Last segment
+                if (offset + memory.Length >= totalLength)
+                {
+                    return memory;
+                }
+
+                offset += memory.Length;
             }
 
-            // We can't just use while(payload.Length > 0) because we need to write a blank final "data: " line
-            // if the payload ends in a newline. For example, consider the following payload:
-            //   "Hello\n"
-            // It needs to be written as:
-            //   data: Hello\r\n
-            //   data: \r\n
-            //   \r\n
-            // Since we slice past the newline when we find it, after writing "Hello" in the previous example, we'll
-            // end up with an empty payload buffer, BUT we need to write it as an empty 'data:' line, so we need
-            // to use a condition that ensure the only time we stop writing is when we write the slice after the final
-            // newline.
+            throw new InvalidOperationException("Could not get last segment from sequence.");
+        }
+
+        private static async Task WriteMessageToMemory(Stream output, ReadOnlySequence<byte> payload)
+        {
             var keepWriting = true;
             while (keepWriting)
             {
-                var span = payload.Span;
-                // Seek to the end of buffer or newline
-                var sliceEnd = span.IndexOf(LineFeed);
-                var nextSliceStart = sliceEnd + 1;
-                if (sliceEnd < 0)
-                {
-                    sliceEnd = payload.Length;
-                    nextSliceStart = sliceEnd + 1;
+                var sliceEnd = payload.PositionOf(LineFeed) ?? payload.End;
 
-                    // This is the last span
-                    keepWriting = false;
-                }
-                if (sliceEnd > 0 && span[sliceEnd - 1] == '\r')
+                ReadOnlySequence<byte> lineSegment;
+                if (sliceEnd == payload.End)
```

Comment:
I don't think this comparison is safe. SequencePosition(s) can't be compared.

/cc @pakrym

Reply:
https://github.com/dotnet/corefx/issues/29001

Reply:
Can you comment on how to achieve this logic without comparing SequencePositions? https://github.com/aspnet/SignalR/blob/1ca2e5f2638f498bb4b57a9d2c69084ecf99bc8f/src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/ServerSentEventsMessageFormatter.cs#L71-L111

Reply:
You can replace first `sliceEnd == payload.End` with `sliceEnd.HasValue`.

Reply:
Here you can just check if `sliceEnd` is `null` rather than using `?? payload.End`